Remote data

RealOrangeOne opened the pull request:
Fixes #476.
Re-implements #465
Extends #379 

Allow loading data from remote endpoints, and parsing through the existing formats `load_data` supports. 

This PR is intentionally designed to reuse as much of the code as possible when reading from data files or remote endpoints. This not only reduces the chance of odd bugs appearing, but means I don't have to write as much! Because I also moved things out into separate files, GitHub makes it looks like i've done way more work than I actually have.

- Support requesting remote data
- Allow loading data as plain text, with no parsing
- Implement a cache to reduce repeat API calls / parsing
- Move data loading into separate file to split it out, it's quite a lot of code!
- Send the current version of zola in the user agent. There's no reason for this, I just thought it might be nice!
- Rename `kind` argument `format`, as it's more appropriate
- If someone tries to load files from outside their site directory, throw an error. This _should_ also include symlinks

It would be possible to edit the headers when sending / receiving a response with the `format` argument. But i decided not to do that, and we / someone else can add it further down the line.

None of this code has been run through `rustfmt`, intentionally.

Note: Currently the test `can_load_remote_data` is failing, when it shouldn't be. This will fix itself once the repo move occurs.

cc @kellpossible, enjoy!

RealOrangeOne committed:
Require EITHER `path` or `url` argument

RealOrangeOne committed:
Refactor global fns into module for future expansion

RealOrangeOne committed:
Abstract load_data to separate file

RealOrangeOne committed:
Change builtin functions to take preloaded data, and lay foundations for
URL type

RealOrangeOne committed:
Automatically determine file kind again if the passed arg is a file

RealOrangeOne committed:
Store path argument as PathBuf

RealOrangeOne committed:
Prevent directory traversal when loading data files

RealOrangeOne committed:
Calculate a key ready for the cache

RealOrangeOne committed:
Implement cache

RealOrangeOne committed:
Request data from remote

RealOrangeOne committed:
Validate URLs

RealOrangeOne committed:
Bake version into user agent

RealOrangeOne committed:
Better handle directory traversal issues, and missing file errors

RealOrangeOne committed:
Update documentation for load_data

RealOrangeOne committed:
Change kind argument to format

RealOrangeOne committed:
Handle 404s better

RealOrangeOne committed:
Restore accidentally commented out constant

Keats commented on components/templates/src/lib.rs:
```diff
@@ -5,6 +5,13 @@ extern crate tera;
+extern crate url;
+
+#[macro_use]
+extern crate clap;
```
why are we loading clap?

RealOrangeOne replied:
See https://github.com/Keats/gutenberg/pull/494#discussion_r227265030

Keats commented on components/templates/Cargo.toml:
```diff
@@ -1,6 +1,7 @@
 [package]
 name = "templates"
-version = "0.1.0"
+# Version must match that of root Cargo.toml
```
why?

RealOrangeOne replied:
See https://github.com/Keats/gutenberg/pull/494#discussion_r227265030

Keats commented on components/templates/Cargo.toml:
```diff
@@ -12,6 +13,11 @@ toml = "0.4"
 csv = "1"
 serde_json = "1.0"
 error-chain = "0.12"
+crypto-hash = "0.3.1"
```
we don't need a crypto hash, we can use the one from std (`std::hash::Hash`) like the imageproc sub-crate does

RealOrangeOne replied:
ah, i somehow missed the implementation for `String` :+1:  And that we don't necessarily need an actual checksum, just some kind of hashed value would do

Keats commented on components/templates/Cargo.toml:
```diff
@@ -12,6 +13,11 @@ toml = "0.4"
+chrono = "0.4"
+reqwest = "0.9"
+url = "1.5"
+clap = "2"
```
I don't think clap is needed?

RealOrangeOne replied:
The reason I loaded `clap` into this package was for the `crate_version` macro for use in the header. Seeing as it's already in the binary, the additional overhead should be minimal.

Keats replied:
I would still remove the version from the header and keep it to just zola. Even if it is already in the bundle, it is pretty suprising to see clap as a dependency there

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+
+enum ProvidedArgument {
+    URL(Url),
+    PATH(PathBuf)
```
enum variants should be lowercase, `Url` and `Path`.
The enum name would be probably clearer as `DataSource` as well

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    if !path_arg.is_some().bitxor(url_arg.is_some()) {
```
I think it would be easier to read as `if path_arg.is_some() && url_arg.is_some()`
The case where both are empty will be taken care of at the end of the function

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+        .map_err(|e| format!("`load_data`: error {} loading file {}", full_path.to_str().unwrap(), e).into());
+}
+
+fn get_output_format_from_args(args: &HashMap<String, Value>, provided_argument: &ProvidedArgument) -> Result<String> {
```
This function should return an enum of the supported types to avoid checking strings later

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+        }
+        return Ok(ProvidedArgument::PATH(full_path));
+    }
+    else if let Some(url) = url_arg {
```
can be only an `if` as all branches in the original `if` return

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+/// Currently the supported formats are json, toml and csv
+pub fn make_load_data(content_path: PathBuf, base_path: PathBuf) -> GlobalFn {
+    let mut headers = header::HeaderMap::new();
+    headers.insert(header::USER_AGENT, format!("{} {}", crate_name!(), crate_version!()).parse().unwrap());
```
is clap added to the dependencies just for that? I would just hardcode `zola` and not have the version imo, it's not that important.

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+        let data = match provided_argument {
+            ProvidedArgument::PATH(path) => read_data_file(&base_path, path),
+            ProvidedArgument::URL(url) => {
+                let mut response = response_client.get(url.as_str()).send().and_then(|res| res.error_for_status()).map_err(|e| format!("Failed to request {}: {}", url, e.status().expect("response status")))?;
```
URL requests should add accept headers to the request since some sites send content following it.

RealOrangeOne replied:
Of course, that makes sense. It doesn't look like there's an existing library to handle converting extensions to mimetypes, but a simple `match` based on the output enum from https://github.com/Keats/gutenberg/pull/494#discussion_r227107910 will cover that easily enough!

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+            "toml" => load_toml(data),
+            "csv" => load_csv(data),
+            "json" => load_json(data),
+            "plain" => to_value(data).map_err(|e| e.into()),
```
I'm not sure about `plain`, what is it even loading?

RealOrangeOne replied:
`plain` is just the format which returns the data raw, with no additional processing. Either the request body or the file contents.

Keats replied:
but when is it needed? I can't really think of a usecase where you would want that

RealOrangeOne replied:
There probably are some use cases out there, and seeing as it was _basically free_ to implement, thought I might as well. It's probably something which may be critical to someone out there

ryanmcgrath replied:
Just a heads up - I actually would make use of `plain` to load CSS and shove it inline at build-time, or to load a base64-encoded value that's ugly in code but needs to exist. They're rare use cases but they do exist.

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,346 @@
+/// tera value
+fn load_json(json_data: String) -> Result<Value> {
+    let json_content = serde_json::from_str(json_data.as_str()).map_err(|e| format!("{:?}", e))?;
+    let tera_value: Value = json_content;
```
you can fold that into the previous line like load_toml I think

RealOrangeOne replied:
I can. I intentionally didn't modify these functions more than I had to, not sure how I missed this line in the cleanup!

Keats commented on components/utils/src/fs.rs:
```diff
@@ -93,6 +100,18 @@ pub fn copy_directory(src: &PathBuf, dest: &PathBuf) -> Result<()> {
 }
 
+pub fn get_file_time(path: &Path) -> Option<SystemTime> {
+    println!("{}", path.display());
```
can remove that

Keats commented on docs/content/documentation/templates/overview.md:
```diff
@@ -186,7 +187,28 @@ template:
+
+#### Data Caching
+
+Data file loading and remote requests are cached in memory memory during build, so multiple requests aren't made to the same endpoint. URLs are cached based on the URL, and data files are cached based on the files modified time. The format is also taken into account when caching, so a request will be sent twice if it's loaded with 2 differnet formats.
```
memory memory in the first line

Keats replied:
differnet -> different in the last words

Keats commented on docs/templates/index.html:
```diff
@@ -14,10 +14,11 @@
 
         <header>
             <nav class="{% block extra_nav_class %}container{% endblock extra_nav_class %}">
+                {% set gh_response = load_data(url="https://api.github.com/repos/getzola/zola", format="json") %}
```
I wouldn't use that in the template of the docs

RealOrangeOne committed:
Remove clap dependency and hardcode package name

RealOrangeOne committed:
Remove unnecessary xnor

RealOrangeOne committed:
Remove unnecessary intermediary variable

RealOrangeOne committed:
Remove debug log

RealOrangeOne committed:
Fix typo in docs

RealOrangeOne committed:
Remove data fetch from docs templates

RealOrangeOne committed:
Rename `ProvidedArgument` enum so it makes more sense

RealOrangeOne committed:
Merge branch 'remote-data' of github.com:RealOrangeOne/gutenberg into remote-data

RealOrangeOne committed:
Use enum for output format!

RealOrangeOne committed:
Use `std::hash::Hash` for hashing rather than MD5

RealOrangeOne committed:
Add accept header to request

RealOrangeOne committed:
impl arg unpack into DataSource enum

Keats commented on Cargo.toml:
```diff
@@ -1,5 +1,6 @@
 [package]
 name = "zola"
+# Update templates Cargo.toml when bumping version
```
can be removed I think

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,403 @@
+enum OutputFormat {
+    TOML,
+    JSON,
+    CSV,
```
I wouldn't uppercase them even if they are acronyms to be consistent with Rust style (https://doc.rust-lang.org/1.0.0/style/style/naming/README.html)

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,403 @@
+    TOML,
+    JSON,
+    CSV,
+    Plain
```
I think I would still remove that for now, it can be added later if necessary

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,403 @@
+}
+
+impl OutputFormat {
+    fn from(output_format: String) -> Result<Self> {
```
you could impl [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html) instead

RealOrangeOne replied:
I did think of that, but because it's possible for it to fail (ie the string doesnt match one of the enum values), it's unsafe, and `TryFrom` is only in nightly right now.

Keats replied:
`FromStr` returns a `Result` though? `FromStr` and `TryFrom` look to be identical for this usecase

RealOrangeOne replied:
Aha, I found my confusion. I was looking at `From` rather than `FromStr`. Yes you're right, `FromStr` would be ideal to implement here!

Keats commented on components/templates/src/global_fns/load_data.rs:
```diff
@@ -0,0 +1,403 @@
+    }
+}
+
+fn get_cache_key(data_source: &DataSource, format: &OutputFormat) -> u64 {
```
Making this part of the impl of DataSource would be cleaner imo

RealOrangeOne replied:
I was toying with where to put it, as it's also equally valid on `OutputFormat`,  but as most of the computation varys for `DataSource`, I guess there makes the most sense of the 2.

Keats commented on docs/content/documentation/templates/overview.md:
````diff
@@ -186,7 +187,28 @@ template:
+{{ response }}
+```
+
+By default, the response body will be returned with no parsing. This can be changed by using the `format` argument as above.
````
Hmm I guess we need `plain` because of that? How about making the `format` argument mandatory?

RealOrangeOne replied:
I could make it mandatory, but the fact it bases it off the file extension is a really nice touch for local data files. Having to specify that my `data.json` file is a JSON file feels unnecessary when it's so easy to guess from the file extension, and change it by default if required.

Keats replied:
I'm not sure myself what is the best choice there. I kind of like the nice UX for ending in JSON but in practice for remote data you will need to specify the format so it feels a bit like it could surprise the end user when something works for local data but not remote.

I would tend towards removing the special casing but not 100% convinced.

Keats reviewed (changes requested):
A couple more comments but looking good!

RealOrangeOne committed:
Remove notice about bumping templates version

RealOrangeOne committed:
Rename enum to match Rust conventions

RealOrangeOne committed:
Don't custom implement FromStr

RealOrangeOne committed:
Move cache key to DataSource, and implement Hash

RealOrangeOne committed:
Fix path tests

We cant use a fixed hash because it's based on the absolute path, which
changes for each dev (probably)

RealOrangeOne committed:
Replace url with one which exists now

RealOrangeOne committed:
Remove unused import

RealOrangeOne committed:
Use httpbin as demp urls

RealOrangeOne committed:
Don't mention path split character in tests

This means tests can run on window / *nix
